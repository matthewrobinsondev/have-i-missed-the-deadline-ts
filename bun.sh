#!/bin/bash

docker compose exec app bun $@