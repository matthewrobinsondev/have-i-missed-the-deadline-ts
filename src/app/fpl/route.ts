export async function GET() {
  const res = await fetch(
    "https://fantasy.premierleague.com/api/bootstrap-static/",
    {
      next: { revalidate: 60 }, // Revalidate every 60 seconds
    },
  );

  const data = await res.json();

  return Response.json(data);
}
