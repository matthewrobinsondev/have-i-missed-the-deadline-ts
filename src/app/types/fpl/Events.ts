export interface Event {
  is_current: boolean;
  is_next: boolean;
  deadline_time: string;
  deadline_time_epoch: number;
  finished: boolean;
}
