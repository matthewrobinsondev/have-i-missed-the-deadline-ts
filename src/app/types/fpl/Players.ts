export interface Player {
  id: number;
  first_name: string;
  second_name: string;
  team: number;
  transfers_in_event: number;
  transfers_out_event: number;
}

export type PlayerTransferKey = "transfers_out_event" | "transfers_in_event";
