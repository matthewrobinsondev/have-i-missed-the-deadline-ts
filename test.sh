#!/bin/bash

docker compose run --rm app bun test $@