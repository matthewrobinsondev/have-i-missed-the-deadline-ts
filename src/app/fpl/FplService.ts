import { log } from "next-axiom";
import { Event } from "../types/fpl/Events";
import { Player, PlayerTransferKey } from "../types/fpl/Players";

export class FplService {
  private fplApi: FplApiInterface;

  constructor(fplApi: FplApiInterface) {
    this.fplApi = fplApi;
  }

  async getDeadline() {
    try {
      const generalInformation = await this.fplApi.getGeneralInformation();

      const currentGameweek = generalInformation.events.find(
        (event: Event) => event.is_current,
      );
      const nextGameweek = generalInformation.events.find(
        (event: Event) => event.is_next,
      );

      return this.calculateDeadline(currentGameweek, nextGameweek);
    } catch (error) {
      if (error instanceof Error) {
        log.error(error.message);
      }

      return 0;
    }
  }

  async getTopTransferredIn() {
    return this.getTopTransferredPlayers("transfers_in_event");
  }

  async getTopTransferredOut() {
    return this.getTopTransferredPlayers("transfers_out_event");
  }

  private async getTopTransferredPlayers(event: PlayerTransferKey) {
    try {
      const generalInformation = await this.fplApi.getGeneralInformation();
      const players = generalInformation.elements;

      players.sort((a: Player, b: Player) => b[event] - a[event]);

      const topTransferredEvents = players
        .slice(0, 10)
        .map((player: Player) => player[event]);

      return players.filter((player: Player) =>
        topTransferredEvents.includes(player[event]),
      );
    } catch (error) {
      if (error instanceof Error) {
        log.error(error.message);
      }
    }
  }

  private calculateDeadline(
    currentGameweek: Event,
    nextGameweek: Event,
  ): number {
    return currentGameweek.finished
      ? Date.parse(nextGameweek.deadline_time)
      : Date.parse(currentGameweek.deadline_time);
  }
}
