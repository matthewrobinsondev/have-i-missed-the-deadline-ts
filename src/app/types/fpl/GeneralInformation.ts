import { Event } from "./Events";
import { Player } from "./Players";

export interface GeneralInformation {
  events: Event[];
  elements: Player[];
}
