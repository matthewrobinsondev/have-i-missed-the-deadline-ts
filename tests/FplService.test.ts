import { FplService } from "../src/app/fpl/FplService";
import { jest, expect, describe, beforeEach, it } from "bun:test";

describe("FplService", () => {
  let fplService: FplService;
  let fplApiMock: any;

  beforeEach(() => {
    fplApiMock = {
      getGeneralInformation: jest.fn(),
    };
    fplService = new FplService(fplApiMock);
  });

  describe("getDeadline", () => {
    it("should return next gameweek deadline if current gameweek is finished", async () => {
      const generalInformation = {
        events: [
          {
            id: 1,
            is_current: true,
            is_next: false,
            finished: true,
            deadline_time: "2023-08-18T17:15:00Z",
            deadline_time_epoch: 1000,
          },
          {
            id: 2,
            is_current: false,
            is_next: true,
            finished: false,
            deadline_time: "2023-08-25T17:15:00Z",
            deadline_time_epoch: 2000,
          },
        ],
      };
      fplApiMock.getGeneralInformation.mockResolvedValueOnce(
        generalInformation
      );

      const result = await fplService.getDeadline();

      // Epoch timestamp for 25th August 17:15:00
      expect(result).toBe(1692983700000);
    });

    it("should return current gameweek deadline if current gameweek is not finished", async () => {
      const generalInformation = {
        events: [
          {
            id: 1,
            is_current: true,
            is_next: false,
            finished: false,
            deadline_time: "2023-08-18T17:15:00Z",
            deadline_time_epoch: 1000,
          },
          {
            id: 2,
            is_current: false,
            is_next: true,
            finished: false,
            deadline_time: "2023-08-25T17:15:00Z",
            deadline_time_epoch: 2000,
          },
        ],
      };
      fplApiMock.getGeneralInformation.mockResolvedValueOnce(
        generalInformation
      );

      const result = await fplService.getDeadline();

      // Epoch timestamp for 18th August 17:15:00
      expect(result).toBe(1692378900000);
    });
  });
});
