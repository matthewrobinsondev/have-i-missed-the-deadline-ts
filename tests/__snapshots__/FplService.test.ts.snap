// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`FplService getTopTransferredIn should return the top 10 transferred in 1`] = `
[
  {
    "first_name": "Player11",
    "id": 111,
    "second_name": "LastName11",
    "team": 211,
    "transfers_in_event": 78811,
    "transfers_out_event": 11411,
  },
  {
    "first_name": "Player15",
    "id": 115,
    "second_name": "LastName15",
    "team": 215,
    "transfers_in_event": 33515,
    "transfers_out_event": 415,
  },
  {
    "first_name": "Player20",
    "id": 120,
    "second_name": "LastName20",
    "team": 220,
    "transfers_in_event": 9320,
    "transfers_out_event": 4290,
  },
  {
    "first_name": "Player23",
    "id": 123,
    "second_name": "LastName23",
    "team": 223,
    "transfers_in_event": 8323,
    "transfers_out_event": 6423,
  },
  {
    "first_name": "Player6",
    "id": 106,
    "second_name": "LastName6",
    "team": 206,
    "transfers_in_event": 8306,
    "transfers_out_event": 7406,
  },
  {
    "first_name": "Player7",
    "id": 107,
    "second_name": "LastName7",
    "team": 207,
    "transfers_in_event": 7307,
    "transfers_out_event": 8407,
  },
  {
    "first_name": "Player22",
    "id": 122,
    "second_name": "LastName22",
    "team": 222,
    "transfers_in_event": 6322,
    "transfers_out_event": 7422,
  },
  {
    "first_name": "Player19",
    "id": 119,
    "second_name": "LastName19",
    "team": 219,
    "transfers_in_event": 6319,
    "transfers_out_event": 419,
  },
  {
    "first_name": "Player5",
    "id": 105,
    "second_name": "LastName5",
    "team": 205,
    "transfers_in_event": 6305,
    "transfers_out_event": 405,
  },
  {
    "first_name": "Player8",
    "id": 108,
    "second_name": "LastName8",
    "team": 208,
    "transfers_in_event": 5308,
    "transfers_out_event": 4085,
  },
]
`;

exports[`FplService getTopTransferredOut should return the top 10 transferred out 1`] = `
[
  {
    "first_name": "Player11",
    "id": 111,
    "second_name": "LastName11",
    "team": 211,
    "transfers_in_event": 78811,
    "transfers_out_event": 11411,
  },
  {
    "first_name": "Player7",
    "id": 107,
    "second_name": "LastName7",
    "team": 207,
    "transfers_in_event": 7307,
    "transfers_out_event": 8407,
  },
  {
    "first_name": "Player22",
    "id": 122,
    "second_name": "LastName22",
    "team": 222,
    "transfers_in_event": 6322,
    "transfers_out_event": 7422,
  },
  {
    "first_name": "Player6",
    "id": 106,
    "second_name": "LastName6",
    "team": 206,
    "transfers_in_event": 8306,
    "transfers_out_event": 7406,
  },
  {
    "first_name": "Player23",
    "id": 123,
    "second_name": "LastName23",
    "team": 223,
    "transfers_in_event": 8323,
    "transfers_out_event": 6423,
  },
  {
    "first_name": "Player21",
    "id": 121,
    "second_name": "LastName21",
    "team": 221,
    "transfers_in_event": 3921,
    "transfers_out_event": 4921,
  },
  {
    "first_name": "Player3",
    "id": 103,
    "second_name": "LastName3",
    "team": 203,
    "transfers_in_event": 3035,
    "transfers_out_event": 4403,
  },
  {
    "first_name": "Player17",
    "id": 117,
    "second_name": "LastName17",
    "team": 217,
    "transfers_in_event": 3157,
    "transfers_out_event": 4317,
  },
  {
    "first_name": "Player9",
    "id": 109,
    "second_name": "LastName9",
    "team": 209,
    "transfers_in_event": 3029,
    "transfers_out_event": 4309,
  },
  {
    "first_name": "Player20",
    "id": 120,
    "second_name": "LastName20",
    "team": 220,
    "transfers_in_event": 9320,
    "transfers_out_event": 4290,
  },
]
`;
